nexus/src/app/crucible.rs

```diff
@@ -262,6 +264,193 @@ impl super::Nexus {
         Ok(returned_region)
     }
 
+    /// Ensure that a running snapshot for a region snapshot exists and is
+    /// running.
+    async fn ensure_crucible_running_snapshot_impl(
+        &self,
+        log: &Logger,
+        dataset: &db::model::Dataset,
+        region_id: Uuid,
+        snapshot_id: Uuid,
+    ) -> Result<(Region, Snapshot, RunningSnapshot), Error> {
+        // Validate with the Crucible agent that both the underlying region and
+        // snapshot exist
+
+        info!(
+            log,
+            "contacting crucible agent to confirm region exists";
+            "dataset" => ?dataset.id(),
+            "region" => ?region_id,
+        );
+
+        let region = match self
+            .maybe_get_crucible_region(log, dataset, region_id)
+            .await?
+        {
+            Some(region) => {
+                info!(
+                    log,
+                    "confirmed the region exists";
+                    "dataset" => ?dataset.id(),
+                    "region" => ?region,
+                );
+
+                region
+            }
+
+            None => {
+                error!(
+                    log,
+                    "region does not exist!";
+                    "dataset" => ?dataset.id(),
+                    "region" => ?region_id,
+                );
+
+                return Err(Error::invalid_request(format!(
+                    "dataset {:?} region {:?} does not exist!",
+                    dataset.id(),
+                    region_id,
+                )));
+            }
+        };
+
+        info!(
+            log,
+            "contacting crucible agent to confirm snapshot exists";
+            "dataset" => ?dataset.id(),
+            "region" => ?region_id,
+            "snapshot" => ?snapshot_id,
+        );
+
```

Comment:
If we just did `maybe_get_crucible_snapshot()`, won't that (by default) also verify
that the region exists as well?  Do we do two laps here so we can log where the
failure is instead of just one call to `maybe_get_crucible_snapshot()`?
(also, less code)

Reply:
If the agent is following its own contract, a snapshot cannot be created without the region being present and in a good state, you're right, but in this function we also return the region that the agent returns us to the caller. This is used in the snapshot create saga to create the mapping from region address to snapshot address.